src/os/bluestore/BlueStore.cc

```diff
@@ -5456,63 +5480,122 @@ int BlueStore::_write_bdev_label(CephContext *cct,
-	 << ": " << cpp_strerror(r) << dendl;
-    goto out;
+  int r = 0;
+  for (uint64_t offset : bdev_label_offsets) {
+    if (alloc_available) {
+      if (offset >= bdev->get_size()) {
+        break;
+      }
+      _unalloc_bdev_label_offset(offset);
+    }
+
+    dout(10) << __func__ << " replicating label in offset=" << offset << dendl;
+
+    r = bl.write_fd(fd, offset);
+    if (r < 0) {
+      derr << __func__ << " failed to write to " << path
+        << ": " << cpp_strerror(r) << dendl;
+      goto out;
+    }
   }
   r = ::fsync(fd);
   if (r < 0) {
     derr << __func__ << " failed to fsync " << path
-	 << ": " << cpp_strerror(r) << dendl;
+      << ": " << cpp_strerror(r) << dendl;
   }
 out:
   VOID_TEMP_FAILURE_RETRY(::close(fd));
   return r;
 }
-
-int BlueStore::_read_bdev_label(CephContext* cct, const string &path,
-				bluestore_bdev_label_t *label)
+int BlueStore::_write_bdev_label(CephContext *cct,
+				 const string &path, bluestore_bdev_label_t label)
 {
-  dout(10) << __func__ << dendl;
-  int fd = TEMP_FAILURE_RETRY(::open(path.c_str(), O_RDONLY|O_CLOEXEC));
+  dout(10) << __func__ << " path " << path << " label " << label << dendl;
+  bufferlist bl;
+  encode(label, bl);
+  uint32_t crc = bl.crc32c(-1);
+  encode(crc, bl);
+  ceph_assert(bl.length() <= BDEV_LABEL_BLOCK_SIZE);
+  bufferptr z(BDEV_LABEL_BLOCK_SIZE - bl.length());
+  z.zero();
+  bl.append(std::move(z));
+
+  int fd = TEMP_FAILURE_RETRY(::open(path.c_str(), O_WRONLY|O_CLOEXEC|O_DIRECT));
   if (fd < 0) {
     fd = -errno;
     derr << __func__ << " failed to open " << path << ": " << cpp_strerror(fd)
 	 << dendl;
     return fd;
   }
-  bufferlist bl;
-  int r = bl.read_fd(fd, BDEV_LABEL_BLOCK_SIZE);
-  VOID_TEMP_FAILURE_RETRY(::close(fd));
+  bl.rebuild_aligned_size_and_memory(BDEV_LABEL_BLOCK_SIZE, BDEV_LABEL_BLOCK_SIZE, IOV_MAX);
+  int r = 0;
+  for (uint64_t offset : bdev_label_offsets) {
+    r = bl.write_fd(fd, offset);
+    if (r < 0) {
+      derr << __func__ << " failed to write to " << path
+        << ": " << cpp_strerror(r) << dendl;
+      goto out;
+    }
+  }
+  r = ::fsync(fd);
   if (r < 0) {
-    derr << __func__ << " failed to read from " << path
-	 << ": " << cpp_strerror(r) << dendl;
-    return r;
+    derr << __func__ << " failed to fsync " << path
+      << ": " << cpp_strerror(r) << dendl;
   }
+out:
+  VOID_TEMP_FAILURE_RETRY(::close(fd));
+  return r;
+}
 
-  uint32_t crc, expected_crc;
-  auto p = bl.cbegin();
-  try {
-    decode(*label, p);
-    bufferlist t;
-    t.substr_of(bl, 0, p.get_off());
-    crc = t.crc32c(-1);
-    decode(expected_crc, p);
-  }
-  catch (ceph::buffer::error& e) {
-    derr << __func__ << " unable to decode label " << path.c_str()
-         << " at offset " << p.get_off()
-	 << ": " << e.what()
-	 << dendl;
-    return -ENOENT;
+int BlueStore::_read_bdev_label(CephContext* cct, const string &path,
+				bluestore_bdev_label_t *label, bool replicated)
```

Comment:
The name 'replicated' is confusing for me.
"false" = require all copies are ok
"true" = be satisfied if one copy is ok
How about "recovery_mode"?